sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/SubscribeTransform.java

```diff
@@ -0,0 +1,128 @@
+    this.options = options;
+  }
+
+  private void checkSubscription(SubscriptionPartition subscriptionPartition) throws ApiException {
+    checkArgument(subscriptionPartition.subscription().equals(options.subscriptionPath()));
+  }
+
+  private Subscriber newSubscriber(Partition partition, Consumer<List<SequencedMessage>> consumer) {
+    try {
+      return options
+          .getSubscriberFactory(partition)
+          .newSubscriber(
+              messages ->
+                  consumer.accept(
+                      messages.stream()
+                          .map(message -> message.toProto())
+                          .collect(Collectors.toList())));
+    } catch (Throwable t) {
+      throw toCanonical(t).underlying;
+    }
+  }
+
+  private PartitionProcessor newPartitionProcessor(
+      SubscriptionPartition subscriptionPartition,
+      RestrictionTracker<OffsetRange, OffsetByteProgress> tracker,
+      OutputReceiver<SequencedMessage> receiver)
+      throws ApiException {
+    checkSubscription(subscriptionPartition);
+    return new PartitionProcessorImpl(
+        tracker,
+        receiver,
+        consumer -> newSubscriber(subscriptionPartition.partition(), consumer),
+        options.flowControlSettings());
+  }
+
+  private RestrictionTracker<OffsetRange, OffsetByteProgress> newRestrictionTracker(
+      SubscriptionPartition subscriptionPartition, OffsetRange initial) {
+    checkSubscription(subscriptionPartition);
+    return new OffsetByteRangeTracker(
+        initial,
+        options.getBacklogReader(subscriptionPartition.partition()),
+        Stopwatch.createUnstarted(),
+        MAX_SLEEP_TIME.multipliedBy(3).dividedBy(4),
+        LongMath.saturatedMultiply(options.flowControlSettings().bytesOutstanding(), 10));
+  }
+
+  private InitialOffsetReader newInitialOffsetReader(SubscriptionPartition subscriptionPartition) {
+    checkSubscription(subscriptionPartition);
+    return options.getInitialOffsetReader(subscriptionPartition.partition());
+  }
+
+  private Committer newCommitter(SubscriptionPartition subscriptionPartition) {
+    checkSubscription(subscriptionPartition);
+    return options.getCommitter(subscriptionPartition.partition());
+  }
+
+  @Override
+  public PCollection<SequencedMessage> expand(PBegin input) {
+    PCollection<SubscriptionPartition> partitions =
+        Create.of(
+                options.partitions().stream()
+                    .map(
+                        partition ->
+                            SubscriptionPartition.of(options.subscriptionPath(), partition))
+                    .collect(Collectors.toList()))
+            .expand(input);
+    // Prevent fusion between Create and read.
+    PCollection<SubscriptionPartition> shuffledPartitions =
+        partitions.apply(Reshuffle.viaRandomKey());
+    return shuffledPartitions.apply(
+        ParDo.of(
+            new PerPartitionSdf(
+                MAX_SLEEP_TIME,
+                this::newInitialOffsetReader,
```

Comment:
Can the `PerPartitionSdf` take the `SubscriberOptions ` as the constructor? Then the `PerParittionSdf` can construct  `Committer`, `PartitionProcessor` and `InitialOffsetReader` by itself, instead of asking the caller to do so?